### DISCUSSION TITLE
fadeToBlackBy and ESP32-HUB75-MatrixPanel-I2S-DMA

### DISCUSSION THREAD

**Discussion by MythicalForce** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#225] [Q&A] [answered]
is it possible to make a `fadeToBlackBy` (as the fastLED lib has) but for `ESP32-HUB75-MatrixPanel-I2S-DMA` ? :S

**Comment by mrcodetastic** [comment]
This library is not a pixel buffer library. It simply drives the Matrix Panel LEDs with the output requested in the code, as quickly as possible. It is not possible to interrogate the value of the pixel colours for further processing once sent to the DMA ring output.

If you want to start doing pixel/screen output manipulation, you'll need to write your own 'screen buffer' class.

Use this example I wrote a while back for a similar questions: https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/discussions/194

**Reply by MythicalForce** [reply]
hmfr... dunno how to use that i looked at it but couldn't get it to work dunno what im doing wrong... i passed it a matrix object and tried drawpixel and then display(); but nothing came up

**Reply by mrcodetastic** [reply]
Put your sketch on git somewhere and I'll take a look for you. It's Christmas afterall.

**Reply by MythicalForce** [reply]
https://github.com/MythicalForce/Temp.git

there you go

**Comment by vortigont** [comment]
@MythicalForce you can use some of the FastLED's functions for this, but it will require to have your own frame buffer in addition to DMA.
You can check [AuroraDemo](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/tree/master/examples/AuroraDemo) on how to work with a frame buffer and this lib., i.e. in file [Effects.h](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/blob/master/examples/AuroraDemo/Effects.h)

```
class Effects {
public:
  CRGB *leds;

  Effects(){
    leds = (CRGB *)malloc(NUM_LEDS * sizeof(CRGB));
}

void DimAll(byte value){
    for (int i = 0; i < NUM_LEDS; i++)
    {
      leds[i].nscale8(value);
    }
 }  
```

**Reply by MythicalForce** [reply]
as mentioned to @mrfaptastic im still unsure how to use it

`
hmfr... dunno how to use that i looked at it but couldn't get it to work dunno what im doing wrong... i passed it a matrix object and tried drawpixel and then display(); but nothing came up
`

**Comment by mrcodetastic** [comment]
Yeah give us a day or so I'll try and come up with a new library example which does what you want. There's been a few queries about such a capability now.

**Comment by mrcodetastic** [comment, marked as answer]
Here's the example implemented:

https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/tree/master/examples/ChainedPanelsScreenBuffer

It shows a downward scrolling 'MythicalForce' text with trailing pixels due to dimAll function (same thing as fadeToBlackBy).

It's built on top of the VirtualMatrix class so it will support chaining of panels etc.

Merry Christmas!

**Comment by tsctrl** [comment]
hi @mrfaptastic, 

i cant relates the way to write own pixel buffer and double buffer approach to solve the flickering issue. 

i understand we can control once buffer is properly populated and flush. but why cant double buffering do the same? detecting if buffer is  full or at least is not empty to avoid flushing the buffer prematurely. 

is good to have custom buffer but i cant think how the flushing custom buffer without some way to check if is the right time to flush end up having the same issue.

whats the issue we having previously which cause double buffer not working correctly on rendering and swapping the buffer? didnt aware on the limitation..

thanks!

**Reply by mrcodetastic** [reply]
This is a pixel buffer for FastLED graphics capability. Allows one to read back the value of a pixel and manipulate. Can't do this with the DMA hardware buffer.

Please keep discussions on 'flicker' to the flicker discussion topic please

**Comment by MythicalForce** [comment]
hmmm is it not possible to Blur only 1 pixel? im trying with the fastLED blur2d but it blurs the entire matrix display! i only want to blur my moving pixel (1 pixel)
what im trying to make here is; i have 1 pixel that spawns from the right and then moves all the way to the left with a pixel trail behind it (kinda like a comet) and i want to "blur" the head/leading pixel

**Reply by mrcodetastic** [reply]
That's already been done in the aurora examples. See how they do it in that code.

**Reply by MythicalForce** [reply]
sadly it hasn`t :/ i've looked through the code! they use the blur2d... it blurs the entire matrix display! not "targeted" pixels

**Reply by mrcodetastic** [reply]
You will have to develop your own targeted blur algo as this library isn't a graphic effects / pixel effect library.
